src/rust/src/lib.rs

```diff
@@ -148,49 +174,354 @@ mod rust_bridge {
+    Ok(RustBuf {
+        data: soroban_test_wasms::CONTRACT_STORAGE
+            .iter()
+            .cloned()
+            .collect(),
+    })
+}
+
+pub(crate) fn get_test_wasm_complex() -> Result<RustBuf, Box<dyn std::error::Error>> {
+    Ok(RustBuf {
+        data: soroban_test_wasms::COMPLEX.iter().cloned().collect(),
+    })
+}
+
+use cxx::CxxVector;
+use cxx::UniquePtr;
+use rust_bridge::CxxBuf;
+use rust_bridge::CxxLedgerInfo;
+use rust_bridge::InvokeHostFunctionOutput;
+use rust_bridge::PreflightCallbacks;
+use rust_bridge::PreflightHostFunctionOutput;
+use rust_bridge::RustBuf;
+use rust_bridge::VersionNumPair;
+use rust_bridge::VersionStringPair;
+use rust_bridge::XDRHashesPair;
 
 mod log;
 use crate::log::init_logging;
 
+fn package_matches_hash(pkg: &cargo_lock::Package, hash: &str) -> bool {
+    // Try comparing hash to hashes in either the package checksum or the source
+    // precise field
+    if let Some(cksum) = &pkg.checksum {
+        if cksum.to_string() == hash {
+            return true;
+        }
+    }
+    if let Some(src) = &pkg.source {
+        if let Some(precise) = src.precise() {
+            if precise == hash {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+// We retain _two copies_ of the soroban_env_host crate, and all of its
+// dependencies, compiled into stellar-core at all times. We refer to these two
+// versions as 'lo' and 'hi' and there is a strict process for upgrading them,
+// described in the document docs/versioning-soroban.md
+
+#[path = "."]
+mod lo {
+    pub(crate) use soroban_env_host_lo as soroban_env_host;
+    pub(crate) const LOGIC_VERSION: u32 = 1;
+    pub(crate) mod contract;
+}
+
+#[path = "."]
+mod hi {
+    pub(crate) use soroban_env_host_hi as soroban_env_host;
+    pub(crate) const LOGIC_VERSION: u32 = super::lo::LOGIC_VERSION + 1;
+    pub(crate) mod contract;
+}
+
+// This function performs a crude dynamic check that the contents of Cargo.lock
+// against-which the current binary was compiled specified _exactly_ the same
+// host dep trees that are stored (redundantly, graphically) in the files
+// host-dep-tree-lo.txt and host-dep-tree.hi, and moreover that the logic
+// version numbers we have assigned in the 'lo' and 'hi' modules below are
+// unchanged from those stored redundantly in the dep tree files.
+//
+// The contents of all 3 files are compiled-in to the binary as static strings.
+// Any discrepancy between the logical content of Cargo.lock and the two derived
+// dep trees will cause the program to abort on startup.
+//
+// The point of this check is twofold: to catch cases where the developer
+// accidentally bumps dependencies (which cargo does fairly easily), and also to
+// make crystal clear when doing a commit that intentionally bumps dependencies
+// which of the two _overlapping dependency trees_ is being affected, and how.
+pub fn check_lockfile_has_expected_dep_trees() {
```

Comment:
nit: this function probably could use a helper function/closure to not duplicate the hi/lo code paths.

Reply:
done